File: backend/app/api/routes/posts.py
from uuid import UUID

from fastapi import APIRouter, File, Form, Query, Response, UploadFile, status

from app.api.deps import CurrentUser, DbSession
from app.schemas.post import FeedResponse, PostPublic, TrendPublic
from app.services import posts as posts_service

router = APIRouter(tags=["posts"])


@router.get("/feed", response_model=FeedResponse, summary="Лента текущего пользователя")
def feed(
    db: DbSession,
    current_user: CurrentUser,
    limit: int = Query(default=20, ge=1, le=50),
    offset: int = Query(default=0, ge=0),
) -> FeedResponse:
    return posts_service.get_feed(db, current_user, limit=limit, offset=offset)


@router.post("", response_model=PostPublic, status_code=status.HTTP_201_CREATED, summary="Создать публикацию")
def create_post(
    db: DbSession,
    current_user: CurrentUser,
    caption: str | None = Form(default=None),
    image: UploadFile = File(...),
) -> PostPublic:
    return posts_service.create_post(db, current_user, caption, image)


@router.get("/trends", response_model=list[TrendPublic], summary="Актуальные хештеги")
def trends(db: DbSession, current_user: CurrentUser, limit: int = Query(default=6, ge=1, le=12)) -> list[TrendPublic]:
    return posts_service.get_trends(db, limit=limit)


@router.get("/{post_id}", response_model=PostPublic, summary="Просмотр публикации")
def get_post(post_id: UUID, db: DbSession, current_user: CurrentUser) -> PostPublic:
    return posts_service.get_post(db, post_id, current_user)


@router.delete("/{post_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Удалить свою публикацию")
def delete_post(post_id: UUID, db: DbSession, current_user: CurrentUser) -> Response:
    posts_service.delete_post(db, post_id, current_user)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{post_id}/like", response_model=PostPublic, summary="Поставить лайк")
def like(post_id: UUID, db: DbSession, current_user: CurrentUser) -> PostPublic:
    return posts_service.like_post(db, post_id, current_user)


@router.delete("/{post_id}/like", response_model=PostPublic, summary="Убрать лайк")
def unlike(post_id: UUID, db: DbSession, current_user: CurrentUser) -> PostPublic:
    return posts_service.unlike_post(db, post_id, current_user)

File: frontend/src/api/posts.ts
import { apiClient, type Comment, type FeedResponse, type Post, type Trend } from './client';

export async function getFeed(limit = 20, offset = 0): Promise<FeedResponse> {
  const { data } = await apiClient.get<FeedResponse>('/feed', { params: { limit, offset } });
  return data;
}

export async function createPost(form: FormData): Promise<Post> {
  const { data } = await apiClient.post<Post>('/posts', form, {
    headers: { 'Content-Type': 'multipart/form-data' }
  });
  return data;
}

export async function getPost(postId: string): Promise<Post> {
  const { data } = await apiClient.get<Post>(`/posts/${postId}`);
  return data;
}

export async function deletePost(postId: string): Promise<void> {
  await apiClient.delete(`/posts/${postId}`);
}

export async function likePost(postId: string): Promise<Post> {
  const { data } = await apiClient.post<Post>(`/posts/${postId}/like`);
  return data;
}

export async function unlikePost(postId: string): Promise<Post> {
  const { data } = await apiClient.delete<Post>(`/posts/${postId}/like`);
  return data;
}

export async function listComments(postId: string): Promise<Comment[]> {
  const { data } = await apiClient.get<Comment[]>(`/posts/${postId}/comments`);
  return data;
}

export async function createComment(postId: string, text: string): Promise<Comment> {
  const { data } = await apiClient.post<Comment>(`/posts/${postId}/comments`, { text });
  return data;
}

export async function getTrends(): Promise<Trend[]> {
  const { data } = await apiClient.get<Trend[]>('/posts/trends');
  return data;
}
